### PR TITLE
WT-11753 Get rid of full object caching in S3 store

### PR DESCRIPTION
**Pull request by sulabhM**
Get rid of the S3 extension's full object local caching. A separate chunk cache is being build into WiredTiger that works across all the storage (extensions).

### CODE REVIEW

File: ext/storage_sources/s3_store/s3_connection.cpp [hunk 1]
```diff
@@ -161,20 +162,27 @@ S3Connection::DeleteObject(const std::string &objectKey) const
...
+    request.SetRange(range);
 
-    // The S3 Object should be downloaded to disk rather than into an in-memory buffer. Use a custom
-    // response stream factory to specify how the response should be downloaded.
-    request.SetResponseStreamFactory([=]() {
-        return (Aws::New<Aws::FStream>(
-          s3AllocationTag, path, std::ios_base::out | std::ios_base::binary));
-    });
+    /*
+     * The requested S3 object's range should be extracted into the given buffer. We create a
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_connection.cpp:178]: a IOstream -> an IOstream
givem -> given

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 2]
```diff
@@ -51,14 +51,15 @@ struct S3FileSystem;
 // Statistics to be collected for the S3 storage.
 struct S3Statistics {
     // Operations using AWS SDK.
-    uint64_t listObjectsCount;  // Number of S3 list objects requests
-    uint64_t putObjectCount;    // Number of S3 put object requests
-    uint64_t getObjectCount;    // Number of S3 get object requests
-    uint64_t objectExistsCount; // Number of S3 object exists requests
+    std::atomic<int64_t> listObjectsCount;         // Number of S3 list objects requests
+    std::atomic<int64_t> putObjectCount;           // Number of S3 put object requests
+    std::atomic<int64_t> objectExistsAndSizeCount; // Number of S3 object exists requests
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:56]: Maybe we can just call it objectExistsCount, I think I renamed it as there was a separate objectExists call different from getting the size, which no longer exists and is reaplaced with read call.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 3]
```diff
@@ -175,120 +146,61 @@ S3SourcePath(const std::string &dir, const std::string &name)
     return (dir + "/" + strippedName);
 }
 
-// Return if the file exists. First checks the cache, and then the S3 Bucket.
+// Return if the file exists, via looking into the S3 Bucket. Also fetches the file size if exists.
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:149]: "Also fetches the file size if exists." -> "Also, if the file exists, fetch the file size."

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 4]
```diff
@@ -315,35 +227,40 @@ S3FileOpen(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name,
...
+        s3->log->LogErrorMessage("S3FileOpen: error checking if the file exists. Ret code:" + ret);
+        return (EINVAL);
+    }
+    if (!objExists) {
+        s3->log->LogErrorMessage(
+          "S3FileOpen: object named " + std::string(name) + " does not exist in the bucket.");
+        return (ENOENT);
+    }
+
+    // Check if there is already an existing file handle open.
```
Comment (sulabhM) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:243]: Ah, we are sharing file handles between threads - generally we do not, as file handles contain the seek pointer. But I suppose it is okay in the cloud cases since file handles do not contain any state saved between the read requests.

File: ext/storage_sources/s3_store/s3_connection.cpp (hunk 1 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_connection.cpp:178]: done [fa8365c](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/fa8365ce9ebfbe01a182b327672b7475457c9f95)

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 2 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_storage_source.cpp]: changed [fa8365c](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/fa8365ce9ebfbe01a182b327672b7475457c9f95)

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 3 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_storage_source.cpp]: changed [fa8365c](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/fa8365ce9ebfbe01a182b327672b7475457c9f95)

File: dist/stat_data.py [hunk 5]
```diff
@@ -123,7 +123,7 @@ class SessionOpStat(Stat):
     def __init__(self, name, desc, flags=''):
         Stat.__init__(self, name, SessionOpStat.prefix, desc, flags)
 class StorageStat(Stat):
-    prefix = 'session'
+    prefix = 'tiered-storage'
```
Comment (korteland) [inline dist/stat_data.py:126]: Oops - good catch!

File: ext/storage_sources/s3_store/s3_connection.cpp (hunk 1 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_connection.cpp:178]: I don't think this is done - that commit looks unrelated.

File: src/conn/conn_stat.c [hunk 6]
```diff
@@ -69,6 +69,8 @@ __wt_conn_stat_init(WT_SESSION_IMPL *session)
 {
     WT_CONNECTION_IMPL *conn;
     WT_CONNECTION_STATS **stats;
+    WT_STORAGE_SOURCE *ss;
+    int64_t ss_stats[7];
```
Comment (korteland) [inline src/conn/conn_stat.c:73]: These look unused?

File: src/include/wiredtiger.in [hunk 7]
```diff
@@ -5217,6 +5217,7 @@ struct __wt_storage_source {
      * the handle will not be subsequently accessed by WiredTiger.
      */
     int (*terminate)(WT_STORAGE_SOURCE *storage_source, WT_SESSION *session);
+
```
Comment (korteland) [inline src/include/wiredtiger.in:5220]: Unnecessary whitespace

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 8]
```diff
@@ -51,14 +51,15 @@ struct S3FileSystem;
 // Statistics to be collected for the S3 storage.
 struct S3Statistics {
     // Operations using AWS SDK.
-    uint64_t listObjectsCount;  // Number of S3 list objects requests
-    uint64_t putObjectCount;    // Number of S3 put object requests
-    uint64_t getObjectCount;    // Number of S3 get object requests
-    uint64_t objectExistsCount; // Number of S3 object exists requests
+    std::atomic<int64_t> listObjectsCount;  // Number of S3 list objects requests
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:54]: Two questions here: (1) was it intentional for these to become signed? And (2) why do these now need to be atomic? (Or - did they always need to be atomic, and this is correcting them?)

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 9]
```diff
@@ -175,120 +146,62 @@ S3SourcePath(const std::string &dir, const std::string &name)
     return (dir + "/" + strippedName);
 }
 
-// Return if the file exists. First checks the cache, and then the S3 Bucket.
+// Return if the file exists, via looking into the S3 Bucket. Also, if the file exists, fetch the
+// file size.
 static int
-S3FileExists(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name, bool *fileExists)
+S3FileExistsAndGetSize(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name,
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:152]: This API seems a little confusing. We still have a `S3FileExists` after this change, so why not have this as `S3FileGetSize`?

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 10]
```diff
@@ -175,120 +146,62 @@ S3SourcePath(const std::string &dir, const std::string &name)
...
+    if (*fileExists && size != NULL)
+        *size = objectSize;
 
-// Checks whether a file corresponding to the provided path exists locally.
-static bool
-LocalFileExists(const std::string &path)
-{
-    std::ifstream f(path);
-    return (f.good());
+    return (ret);
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:175]: This works in kind of a strange way - it seems we can return non-zero if the file doesn't exist. But it's not an error for the file not to exist (since we presumably don't know, and that's why we're checking.)

What about having this return 0 in all non-error cases, and forcing the caller to rely on `*found`?

If we just make this `S3FileGetSize`, then we can return "not found" as an error, and `S3FileExists` can look specifically for that "not found". (And return any other error.)

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 11]
```diff
@@ -175,120 +146,62 @@ S3SourcePath(const std::string &dir, const std::string &name)
...
 static int
 S3FileClose(WT_FILE_HANDLE *fileHandle, WT_SESSION *session)
 {
-    int ret = 0;
     S3FileHandle *s3FileHandle = (S3FileHandle *)fileHandle;
     S3Storage *s3 = s3FileHandle->storage;
-    WT_FILE_HANDLE *wtFileHandle = s3FileHandle->wtFileHandle;
+
+    // If there are other active instances of the file being open, do not close file handle.
+    if (--s3FileHandle->referenceCount != 0)
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:192]: Are file handles shared between threads? This reference count might need to be an atomic.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 12]
```diff
@@ -315,35 +228,40 @@ S3FileOpen(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name,
         return (EINVAL);
     }
 
+    // Check if the file exists and get the size
+    bool objExists;
+    wt_off_t objSize;
+    if ((ret = S3FileExistsAndGetSize(fileSystem, session, name, &objExists, &objSize)) != 0) {
+        s3->log->LogErrorMessage("S3FileOpen: error checking if the file exists. Ret code:" + ret);
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:235]: `strerror` is a more useful tool here - the numeric code can change between systems so it's better to print some text.

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 13]
```diff
@@ -828,38 +716,7 @@ static int
...
-    std::filesystem::permissions(destPath.c_str(),
-      std::filesystem::perms::owner_read | std::filesystem::perms::group_read |
-        std::filesystem::perms::others_read,
-      std::filesystem::perm_options::add, ec);
-    ret = ec.value();
-    if (ret != 0) {
-        ret = errno;
-        s3->log->LogErrorMessage("S3FlushFinish: read permissions of " + destPath + " failed");
-    }
-    return (ret);
```
Comment (korteland) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:862]: I think I'm missing something - was this all for the caching subsystem? It's a little surprising to me that this function can now be a no-op.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 8 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:54]: They are intentionally signed, as they will eventually map to WiredTiger stats, which are also signed. Also, yes, they needed to be atomic as they get incremented and accessed concurrently. We are correcting them now.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 9 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:152]: I admit I dont like the name "S3FileExistsAndGetSize", I also did not think hard there. S3FileExists calls S3FileExistsAndGetSize() to check if the file exists, there is a different call that calls this function to get the size. It is sort of an internal function to the two filesystem APIs that we support. Please do suggest better naming.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 10 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:175]: Agreed, the semantics of this function could be cleaned further.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 13 above)
Comment (sulabhM) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:862]: Ah yes, overall, throughout the tiered storage changes, there is a realisation that a flush finish might not be needed at all. It was meant to be called post flushing, but there is nothing really to do.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 8 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:54]: Got it - thanks.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 13 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:862]: OK, that makes sense. I assume we still need to provide something here since it's a callback? If so, maybe we could have a ticket to remove that callback from the API. Doing that removal feels out of scope for this ticket.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 9 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:152]: I would just have `S3FileGetSize` and return `ENOENT` if it doesn't exist - assuming the S3 API gives us some way to tell what the problem is. (It also might just return the right code, in which case this all gets simpler.)

File: ext/storage_sources/s3_store/s3_connection.cpp (hunk 1 above)
Comment (ruby-oujo) [reply ext/storage_sources/s3_store/s3_connection.cpp:178]: NP grammar while these changes are still open: "can be downloaded directly"

File: test/suite/test_tiered06.py [hunk 14]
```diff
@@ -315,11 +311,6 @@ def test_ss_file_systems(self):
```
Comment (ruby-oujo) [inline test/suite/test_tiered06.py:311]: the line above: `  bad_cachedir = '/BAD'` should be deleted - it is now unused.

**Comment by ruby-oujo** [comment]
I took a review of the whole file - commenting here as I am unable to comment on those specific lines. 
A few redundant cache related pieces are still left in. 

For example the whole function 
`
    # Check that objects are in the cache directory after flush_finish.
    def check_caches(self, expect1, expect2):
        got = sorted(list(os.listdir(self.cachedir1)))
        expect = sorted(self.suffix(expect1, 'wtobj'))
        self.assertEquals(got, expect)
        got = sorted(list(os.listdir(self.cachedir2)))
        expect = sorted(self.suffix(expect2, 'wtobj'))
        self.assertEquals(got, expect)`

File: test/suite/helper_tiered.py [hunk 15]
```diff
@@ -140,7 +140,6 @@ def gen_tiered_storage_sources(random_prefix='', test_name='', tiered_only=False
         ('s3_store', dict(is_tiered = True,
```
Comment (ruby-oujo) [inline test/suite/helper_tiered.py:140]: Line 110: Left an instance of has_cache
`            has_cache = True,`

File: test/suite/test_tiered12.py [hunk 16]
```diff
@@ -91,13 +91,6 @@ def test_tiered(self):
 
         # Sleep more than the one second stress timing amount and give the thread time to run.
         time.sleep(2)
-        # After sleeping, the internal thread should have created the cached object.
```
Comment (ruby-oujo) [inline test/suite/test_tiered12.py:94]: This entire section is redundant now from above 

>         # Default cache location is cache-<bucket-name>
>         cache = "cache-" + self.bucket
>         # The bucket format for the S3 store is the name and the region separataed by a semi-colon.
>         # Strip off the region to get the cache folder.
>         if self.ss_name == 's3_store':
>             cache = cache[:cache.find(';')]

File: test/suite/test_tiered12.py [hunk 17]
```diff
@@ -91,13 +91,6 @@ def test_tiered(self):
...
-        if self.has_cache:
```
Comment (ruby-oujo) [inline test/suite/test_tiered12.py:95]: Also the comment above line 70 needs to be edited. 

>         # So after flush tier completes, check that the cached object does not
>         # exist. Then sleep and check that it does exist.
>         #
>

File: ext/storage_sources/s3_store/s3_storage_source.cpp [hunk 18]
```diff
@@ -31,8 +31,8 @@
 #include <fstream>
 #include <list>
 #include <errno.h>
```
Comment (ruby-oujo) [inline ext/storage_sources/s3_store/s3_storage_source.cpp:33]: this looks unused `#include <errno.h>`

File: test/suite/test_tiered12.py (hunk 17 above)
Comment (jiechenbo) [reply test/suite/test_tiered12.py:95]: removed [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: test/suite/test_tiered12.py (hunk 16 above)
Comment (jiechenbo) [reply test/suite/test_tiered12.py:94]: removed [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: test/suite/helper_tiered.py (hunk 15 above)
Comment (jiechenbo) [reply test/suite/helper_tiered.py:139]: removed [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: test/suite/test_tiered06.py (hunk 14 above)
Comment (jiechenbo) [reply test/suite/test_tiered06.py:301]: removed [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 12 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_storage_source.cpp]: Thank you changed here [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 11 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:200]: I have changed it to an atomic [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)

File: ext/storage_sources/s3_store/s3_connection.cpp (hunk 1 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_connection.cpp]: oops thank you! i have fixed it up now.

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 13 above)
Comment (jiechenbo) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:862]: Yes, I believe this has been mentioned already when finishing the azure and gcp project.

**Review by ruby-oujo — APPROVED** [review]
LGTM

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 12 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_storage_source.cpp]: I still have a nitpick with the new code sorry. There should be a space before the string, and it's no longer a code getting printed, so it would be a little clearer without that bit of the text (i.e. `S3FileOpen: error checking if the file exists: " + std::string(...)`

File: ext/storage_sources/s3_store/s3_storage_source.cpp (hunk 10 above)
Comment (korteland) [reply ext/storage_sources/s3_store/s3_storage_source.cpp:168]: Thanks, it reads a lot more clearly now!

**Review by korteland — APPROVED** [review]
LGTM! Just the error string nitpick, no need for a re-review.

**Review by korteland — APPROVED** [review]
LGTM!

**Comment by jiechenbo** [comment]
evergreen merge
